src/modal/modal_ref.ts

```diff
@@ -0,0 +1,45 @@
+import {ComponentRef, ViewContainerRef} from '@angular/core';
+
+import {NgbModalBackdrop} from './modal_backdrop';
+import {NgbModalWindow} from './modal_window';
+
+export class NgbModalRef {
+  private resolve: (result: any) => void;
+  private reject: (reason: any) => void;
+  result: Promise<any>;
+
+  constructor(
+      private _viewContainerRef: ViewContainerRef, private _windowCmptRef: ComponentRef<NgbModalWindow>,
+      private _backdropCmptRef?: ComponentRef<NgbModalBackdrop>) {
+    _windowCmptRef.instance.dismissEvent.subscribe((reason: any) => { this.dismiss(reason); });
+
+    this.result = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+  }
+
+  close(result: any) {
+    if (this._windowCmptRef) {
+      this.resolve(result);
+      this.removeModalElements();
+    }
+  }
+
+  dismiss(reason: any) {
+    if (this._windowCmptRef) {
+      this.reject(reason);
+      this.removeModalElements();
+    }
+  }
+
+  private removeModalElements() {
+    this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._windowCmptRef.hostView));
+    if (this._backdropCmptRef) {
+      this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._backdropCmptRef.hostView));
```

Comment:
This is only temporally here till we add support for stacked modals (if we decide to do so!). But yes, I agree that we will have to move this logic if we decide to support stacked modals.